### ISSUE TITLE
OpenNMT benchmarking versus other pytorch libraries. 

### ISSUE THREAD

**Issue by PetrochukM** [OpenNMT/OpenNMT-py#139]
There is certainly a lot of great work going into OpenNMT. But with every new feature added from some paper, do we have any sense of those features in tandem helping OpenNMT?

This gist compares OpenNMT-py v.s. pytorch-seq2seq on newstest2013.

pytorch-seq2seq has a simple LSTM + Attention model and it achieves:
```
Finished epoch 50, Dev Perplexity: 1695.8370
```

While OpenNMT is much more mature library does worse:
```
Epoch 50,    40/   47; acc:   7.64; ppl: 1500.51;12384 src tok/s; 13280 tgt tok/s;    246 s elapsed
Train perplexity: 1546
Train accuracy: 6.97936
Validation perplexity: 24962.4
Validation accuracy: 4.8146
Decaying learning rate to 1.77636e-15
```

What are the goals of this library?

**Comment by yoonkim** [comment]
With such a small dataset the comparison is not really meaningful.

I would say you would need a dataset of at least 100K sentences to get any indication of performance (this is for translation--for other tasks you may be able to get away with less).

**Comment by srush** [comment]
Thank you for the comparison though. We will set up as benchmark to make sure it does at least this well. 

We have been using this data for comparison. It is about 200k sentences. 

https://github.com/OpenNMT/IntegrationTesting/tree/master/data

Would you be willing to run the pytorch-seq2seq model for comparison sake and post the ppl?

**Comment by Deepblue129** [comment]
@kylegao91

**Comment by kylegao91** [comment]
@Deepblue129 I agree that the newstest2013 is too small for benchmarking the effectiveness.  I used it mainly for evaluating the efficiency.  
@srush I will run experiments on larger dataset once the speed issue is resolved.

**Comment by srush** [comment]
Any update on this? Would love to post these benchmarks now that our code is more stable.

**Comment by dalegebit** [comment]
I have similar problems repeating results with PyOpenNMT. I train my English-to-German nmt model with Europarl v7, and almost completely copy the parameter setting and corpus mentioned in [I](http://opennmt.net/Models/) (but I didn't use preprocess.lua and aggressive tokenizer, which are also mentioned). Even with many tries, I cannot reduce the perplexity below 15 in validation set. I wonder why this happens? Have you tested PyOpenNMT on Europarl v7 and achieved 7.19 PPL in newstest2013.deen? Would mind you releasing the benchmarks that we can refer to?

**Comment by srush** [comment]
We haven't tried this yet, but we have replicated other similar benchmarks,
so I am surprised it is so far off. Can you send over your logs and
command?

**Comment by dalegebit** [comment]
Sorry, I missed your message. Sure. Here are my logs, adam version and sgd version. Adam version has larger ppl on validation set but better accuracy, while sgd version has smaller ppl but worse accuracy. But none of them can achieve <15 ppl on validation set. In the logs, I only show the result of 13 epoches, however, even if I run many more epoches, there is no sign that they could reach that goal. I only used default parameters and default learning rate scheduling, except for changing the initial learning rate and adding `-input 1` and `-extra_shuffle`.
https://github.com/dalegebit/OpenNMT-py/blob/multi-gpu/log/benchmarks_opitm-adam_lr-0.001.txt
https://github.com/dalegebit/OpenNMT-py/blob/multi-gpu/log/benchmarks_optim-sgd_lr-0.01.txt

**Comment by kylegao91** [comment]
Here is the log of latest (0.1.4) pytorch-seq2seq running on your integration test data with its default setting:
https://gist.github.com/kylegao91/91534e220cf745dd976ad24bb4bac4c0

Look like `pytorch-seq2seq` still has a lot of room to improve both speed and accuracy.  I post the log here for discussions, but, for PR reasons, please do not include it if you are going to publish a benchmark.

**Comment by vince62s** [comment]
closing this, quite old.
We reproduce SOA results with the transformer. (comparable to the original paper)
